### PR TITLE
Update codegen module to generate service waiter implementation classes

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
Update codegen module to generate service waiter implementation classes
Note that it currently only has one `WaiterAcceptor` and we will add more once waiterAcceptor generation part is done.
More waiter tests will be added as well after we have the complete waiter code

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    @Override
+    protected Optional<String> additionalWaiterConfig() {
+        return Optional.of(".scheduledExecutorService(executorService)");
+    }
+
+    @Override
+    protected void additionalConstructorInitialization(MethodSpec.Builder method) {
+        method.beginControlFlow("if (builder.executorService == null)")
+              .addStatement("this.executorService = $T.newScheduledThreadPool(5, new $T().threadNamePrefix"
+                            + "(\"waiters-ScheduledExecutor\").build())",
```
Comment (dagnir) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java:88]: minor: use `$S` here so you don't need to escape the quotes

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java [hunk 2]
```diff
@@ -35,11 +37,23 @@
...
+         * Sets a custom {@link ScheduledExecutorService} that will be used to schedule async polling attempts
+         * <p>
+         * This executorService must be closed by the caller when it is ready to be disposed. The SDK will not close the
+         * executorService when the waiter is closed
+         *
+         * @param executorService
+         *        the executorService to set
+         * @return a reference to this object so that method calls can be chained together.
+         */
+        Builder executorService(ScheduledExecutorService executorService);
```
Comment (dagnir) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java:55]: Should we have provider overload, where the returned executors are managed by the SDK?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @Override
+    protected Optional<String> additionalWaiterConfig() {
+        return Optional.of(".scheduledExecutorService(executorService)");
+    }
+
+    @Override
+    protected void additionalConstructorInitialization(MethodSpec.Builder method) {
+        method.beginControlFlow("if (builder.executorService == null)")
```
Comment (dagnir) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java:86]: Hmm why does the async waiter need a threadpool? Would it be enough to just rely on the async client?

Comment (zoewangg) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java:86]: We still need a `ScheduledExecutorService` to schedule the attempts.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java (hunk 2 above)
Comment (zoewangg) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java:55]: I like the idea, but we don't have it for the futureCompletionExecutorService override either. I feel this is additive and we can add it in the future if people are asking for it. I'd expect most customers use the waiters retrieved the service clients, where they don't need to configure it at all.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java (hunk 3 above)
Comment (dagnir) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java:86]: Ah gotcha forgot about the backoffs

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java (hunk 2 above)
Comment (dagnir) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/waiters/query-async-waiter-interface.java:55]: Sounds good

**Review by dagnir — APPROVED** [review]

**Pull request merged**